[Ready for review] use gfile to support remote directories

f4hy opened the pull request:
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## What does this PR do?
Fixes #2161 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

f4hy commented:
Thanks @samj1912 fix the one missing call. I also cleaned up how this was being called and fixed up the tests. The tests were using the tmpdir pytest fixture which gives a `py.path.local` object where the contract in many places is type annotated `str` or `Optional[str]`. So I felt the best choice was to update the tests. Howerver it might be a better solution to just make the contracts some sort of path like object.

f4hy commented:
Rebased since the requirements paths changed on master.

f4hy commented:
I will leave this a draft PR in case a different solution is suggested for issue #2161

f4hy commented:
Hmm. ok the tests here are now failing on just the 'minimal' builds where it passes on the lastest. A better approach here might be to wrap all the IO in something that can detect what is installed. If running on a minimum of dependencies only support local disk writes, and if more/later deps are installed support cloud file paths.

Borda commented on tests/trainer/test_trainer.py:
```diff
@@ -501,7 +501,7 @@ def test_benchmark_option(tmpdir):
     # fit model
     trainer = Trainer(
-        default_root_dir=tmpdir,
+        default_root_dir=str(tmpdir),
```
rather so this casting in Trainer

f4hy replied:
Sounds good. will make that change.

f4hy replied:
The reason to do it here is that the trainer and other places _correctly_ type annotate that it should be a `str` or `Optional[str]` and it is hte tests which are wrong in passing a path object. @Borda I think if the tests are going to do this we should update the type annotations to accept a union of string and path. what do you think?

Borda replied:
yes, then pls update annotation :]

pep8speaks commented:
Hello @f4hy! Thanks for updating this PR. 













* In the file [`pytorch_lightning/utilities/cloud_io.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/bdd67e5983c66330640499c82e1af24524154394/pytorch_lightning/utilities/cloud_io.py):

> [Line 33:1](https://github.com/PyTorchLightning/pytorch-lightning/blob/bdd67e5983c66330640499c82e1af24524154394/pytorch_lightning/utilities/cloud_io.py#L33): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

williamFalcon commented:
@f4hy is this ready for 0.8.2 today?

f4hy commented:
> @f4hy is this ready for 0.8.2 today?

Yep! Just removed some extra cruft. If you @williamFalcon think this is a good solution to this then I think its ready to go now. I was worried the tests are not passing on the 'minimal' builds, I tried to fix it but looks like its failing on master for minimal right now.

Borda commented:
btw, how does it goes with #2175?

f4hy commented:
> btw, how does it goes with #2175?

So s3 support is just one cloud hosting provider. My pr should support any of the backends that tensorboard currently supports which is `s3/hdfs/gcs/local` and maybe others have been added. So if #2175 is merged it wont solve the issue I have of #2161 . 

If the other PR does something this one does not we can add it here, but my understanding is that this is a superset of that.

Borda commented:
@f4hy ho, how is it going, can we get this done in next release?

f4hy commented:
@Borda Hey ya looks like I need to rebase an test again now. Let me get that done tomorrow. Hopefuly we can get this in soon.

f4hy commented:
@Borda I have rebased master and updated the tests so now all pass except against windows. I am not quite sure how to test against windows and some tips would be welcome. I believe it is just due to line-ending issues. 

On a larger issue though, is this design acceptable? I know others have tried to solve similar problems. My issue is my training happens on a Kubernetes cluster where I only have remote file systems to store data so I want all logs/checkpoints/outputs to be written to remote/cloud storage. I think this is a reasonable way to approach it since tensorboard has this capability and is already a dependency but other approaches can work as well.

f4hy commented:
@Borda Rebased again. Now passes all tests. Should be ready for review.

Let me know any ideas on how to improve this. I would like better ways of testing it. The concern I have is ensuring we don't later add some methods to write files that only work locally. So somehow testing this can actually work with remote files would be good, but not sure how to do that without something like docker compose to set that up. Ideas welcome.

igorgad commented:
Hi, @f4hy. Was this already merged? I cannot see it on release 1.1.2. Did it got removed somehow?

f4hy commented:
We went with fsspec instead. See the other PRs by me. Remote directories should be supported but with fsspec instead
